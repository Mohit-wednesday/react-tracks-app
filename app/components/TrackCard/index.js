/* eslint-disable no-console */
/**
 *
 * TrackCard
 *
 */

import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Card, Button } from 'antd';
import { colors } from '@themes';
import If from '@components/If';
import { PlayCircleOutlined, PauseCircleOutlined } from '@ant-design/icons';

const { Meta } = Card;

const CustomCard = styled(Card)`
  & > div {
    display: flex;
  }
  && {
    width: 22rem;
    margin: 1rem 0;
    height: ${(props) => (props.height ? props.height : 10)}rem;
    background-color: ${colors.trackCardColor};
  }
`;

const Text = styled(Meta)`
  && {
    margin: 0.3rem;
    width: 15em;
  }
`;

const Wrapper = styled.div`
  && {
    padding: 0 1em;
  }
`;

const Image = styled.img`
  && {
    width: 5rem;
    height: ${(props) => (props.height ? props.height : 6)}rem;
  }
`;
const CustomButton = styled(Button)`
  display: flex !important;
  justify-content: center !important;
  align-items: center !important;
  border: none !important;
`;

const CustomPlay = styled(PlayCircleOutlined)`
  font-size: 2rem !important;
  cursor: pointer;
`;
const CustomPause = styled(PauseCircleOutlined)`
  font-size: 2rem !important;
  cursor: pointer;
`;

function TrackCard({ track, handleOnClick }) {
  const { trackName, artistName, artworkUrl100, previewUrl, trackId } = track;

  const audioElement = useRef();
  const [play, setPlay] = useState(false);

  const handleAudio = (event, url) => {
    event.preventDefault();

    const isPaused = audioElement.current.paused;
    console.log(isPaused);
    if (isPaused) {
      audioElement.current.src = url;
      audioElement.current.play();
    } else {
      audioElement.current.pause();
    }
    setPlay(!play);
    if (handleOnClick) {
      handleOnClick(audioElement);
    }
  };

  return (
    <Link to={`/track/${trackId}`}>
      <CustomCard data-testid="track-card">
        <Image src={artworkUrl100} />
        <Wrapper>
          <Text title={trackName} description={artistName} />
          <CustomButton
            shape="circle"
            onClick={(e) => handleAudio(e, previewUrl)}
            icon={
              <If condition={!audioElement.current?.paused && audioElement.current?.src} otherwise={<CustomPlay />}>
                <CustomPause />
              </If>
            }
          />
        </Wrapper>
        <audio ref={audioElement}></audio>
      </CustomCard>
    </Link>
  );
}

TrackCard.propTypes = {
  handleOnClick: PropTypes.func,
  track: PropTypes.shape({
    trackName: PropTypes.string,
    artistName: PropTypes.string,
    artworkUrl100: PropTypes.string,
    previewUrl: PropTypes.string,
    trackId: PropTypes.number
  })
};

export default TrackCard;
