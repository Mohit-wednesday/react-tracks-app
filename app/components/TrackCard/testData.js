const testData = {
  wrapperType: 'track',
  kind: 'song',
  artistId: 966309175,
  collectionId: 1445949265,
  trackId: 1445949267,
  artistName: 'Post Malone & Swae Lee',
  collectionName: 'Spider-Man: Into the Spider-Verse (Soundtrack From & Inspired by the Motion Picture)',
  trackName: 'Sunflower',
  collectionCensoredName: 'Spider-Man: Into the Spider-Verse (Soundtrack From & Inspired by the Motion Picture)',
  trackCensoredName: 'Sunflower (Spider-Man: Into the Spider-Verse)',
  collectionArtistId: 36270,
  collectionArtistName: 'Various Artists',
  artistViewUrl: 'https://music.apple.com/us/artist/post-malone/966309175?uo=4',
  collectionViewUrl:
    'https://music.apple.com/us/album/sunflower-spider-man-into-the-spider-verse/1445949265?i=1445949267&uo=4',
  trackViewUrl:
    'https://music.apple.com/us/album/sunflower-spider-man-into-the-spider-verse/1445949265?i=1445949267&uo=4',
  previewUrl:
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/c4/1d/26/c41d2673-dd3f-f320-938f-79be41ae3841/mzaf_7670720480871265544.plus.aac.p.m4a',
  artworkUrl30:
    'https://is2-ssl.mzstatic.com/image/thumb/Music125/v4/4b/30/2c/4b302cb6-7a14-5464-4e97-0577e9d0be49/18UMGIM82277.rgb.jpg/30x30bb.jpg',
  artworkUrl60:
    'https://is2-ssl.mzstatic.com/image/thumb/Music125/v4/4b/30/2c/4b302cb6-7a14-5464-4e97-0577e9d0be49/18UMGIM82277.rgb.jpg/60x60bb.jpg',
  artworkUrl100:
    'https://is2-ssl.mzstatic.com/image/thumb/Music125/v4/4b/30/2c/4b302cb6-7a14-5464-4e97-0577e9d0be49/18UMGIM82277.rgb.jpg/100x100bb.jpg',
  collectionPrice: 11.99,
  trackPrice: 1.29,
  releaseDate: '2018-10-18T12:00:00Z',
  collectionExplicitness: 'notExplicit',
  trackExplicitness: 'notExplicit',
  discCount: 1,
  discNumber: 1,
  trackCount: 13,
  trackNumber: 2,
  trackTimeMillis: 158040,
  country: 'USA',
  currency: 'USD',
  primaryGenreName: 'Hip-Hop/Rap',
  isStreamable: true
};
export default testData;
