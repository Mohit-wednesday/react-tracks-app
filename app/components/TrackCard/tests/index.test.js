/* eslint-disable no-console */
/* eslint-disable no-unused-vars */
/**
 *
 * Tests for TrackCard
 *
 */

import React from 'react';
import { fireEvent, waitFor } from '@testing-library/dom';
import { renderProvider, timeout, createSpyOnAudio } from '@utils/testUtils';
import TrackCard from '../index';
import history from '@app/utils/history';
import testData from '../testData';
import { translate } from '@app/components/IntlGlobalProvider/index';

describe('<TrackCard />', () => {
  let handleOnClick = jest.fn();
  let playSpy;
  let pauseSpy;

  beforeAll(() => {
    playSpy = createSpyOnAudio('play');
    pauseSpy = createSpyOnAudio('pause');
  });

  beforeEach(() => {
    handleOnClick = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const MockTrackCard = () => {
    return <TrackCard {...testData} handleOnClick={handleOnClick} />;
  };

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<MockTrackCard />, history);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackCard component', () => {
    const { getAllByTestId } = renderProvider(<MockTrackCard />, history);
    expect(getAllByTestId('track-card').length).toBe(1);
  });

  it('should render trackName and artistName', () => {
    const { getByTestId } = renderProvider(<MockTrackCard />, history);
    expect(getByTestId('artist-name')).toHaveTextContent(testData.artistName);
    expect(getByTestId('track-name')).toHaveTextContent(testData.trackName);
  });

  it('should call handleOnClick and handleAudio callbacks accordingly', async () => {
    let audio;
    let handleAudioSpy = jest.fn();
    const { getByTestId } = renderProvider(<MockTrackCard />, history);
    audio = getByTestId('audio');
    fireEvent.click(getByTestId('track-control-button'), { onClick: handleAudioSpy() });
    await timeout(500);

    expect(handleOnClick).toBeCalledWith({ current: audio });

    fireEvent.click(getByTestId('track-control-button'), { onClick: handleAudioSpy() });
    await timeout(500);

    expect(handleAudioSpy).toBeCalled();
    expect(handleOnClick).toBeCalledWith({ current: audio });
  });

  it('should play the audio if the play pause button clicked', async () => {
    const { getByTestId } = renderProvider(<MockTrackCard />, history);
    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
  });

  it('should pause the playing audio if the play pause button clicked', async () => {
    const { getByTestId } = renderProvider(<MockTrackCard />, history);
    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));

    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(pauseSpy).toHaveBeenCalledTimes(1));
  });
  it('should render the fallback text messages if data is empty', () => {
    const { getByTestId } = renderProvider(<TrackCard />);
    expect(getByTestId('track_name_unavailable')).toHaveTextContent(translate('track_name_unavailable'));
    expect(getByTestId('artist_name_unavailable')).toHaveTextContent(translate('artist_name_unavailable'));
  });
});
