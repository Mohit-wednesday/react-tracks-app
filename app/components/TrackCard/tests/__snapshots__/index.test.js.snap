// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <a
      href="/track/1445949267"
    >
      <div
        class="ant-card ant-card-bordered TrackCard__CustomCard-sc-10bk196-0 fYvImm"
        data-testid="track-card"
        shadow="0 0 10px 1px #D3D3D3"
      >
        <div
          class="ant-card-body"
        >
          <img
            class="TrackCard__Image-sc-10bk196-3 cVQUAq"
            src="https://is2-ssl.mzstatic.com/image/thumb/Music125/v4/4b/30/2c/4b302cb6-7a14-5464-4e97-0577e9d0be49/18UMGIM82277.rgb.jpg/100x100bb.jpg"
          />
          <div
            class="TrackCard__Wrapper-sc-10bk196-2 jYEldH"
          >
            <p
              class="T__StyledText-gjlic1-0 kYcyRq TrackCard__Text-sc-10bk196-1 hQCbRo"
              data-testid="track-name"
              title="Sunflower"
            >
              Sunflower
            </p>
            <p
              class="T__StyledText-gjlic1-0 kYcyRq TrackCard__Text-sc-10bk196-1 hQCbRo"
              data-testid="artist-name"
              title="Post Malone & Swae Lee"
            >
              Post Malone & Swae Lee
            </p>
            <button
              class="ant-btn ant-btn-icon-only TrackCard__CustomButton-sc-10bk196-4 gFdyeI"
              data-testid="track-control-button"
              padding="0.2"
              type="button"
            >
              <span
                aria-label="play-circle"
                class="anticon anticon-play-circle"
                role="img"
                style="font-size: 2rem;"
              >
                <svg
                  aria-hidden="true"
                  data-icon="play-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M719.4 499.1l-296.1-215A15.9 15.9 0 00398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 000-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <audio
            class="TrackCard__StyledAudio-sc-10bk196-5 etPQVC"
            data-testid="audio"
            src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/c4/1d/26/c41d2673-dd3f-f320-938f-79be41ae3841/mzaf_7670720480871265544.plus.aac.p.m4a"
          />
        </div>
      </div>
    </a>
  </div>
</body>
`;
